hibernate-core/src/main/java/org/hibernate/loader/ast/internal/LoaderHelper.java

```diff
@@ -90,6 +90,9 @@ public static void upgradeLock(Object object, EntityEntry entry, LockOptions loc
 					entry.forceLocked( object, nextVersion );
 				}
 				else {
+					if ( !entry.isExistsInDatabase() ) {
+						session.forceFlush( entry );
+					}
 					persister.lock( entry.getId(), entry.getVersion(), object, lockOptions, session );
```

Comment:
Maybe I'm missing something, but if the `forceFlush` succeeds i.e. does an insert that doesn't fail, it is guaranteed that this transaction by definition already has an exclusive lock on the inserted row, so AFAICT, there is no need to run the locking select.

```suggestion
					if ( entry.isExistsInDatabase() ) {
						persister.lock( entry.getId(), entry.getVersion(), object, lockOptions, session );
					}
					else {
						session.forceFlush( entry );
					}
```